protocol/exporting-action-tests.md

```diff
@@ -0,0 +1,151 @@
+
+# Proposed Solution
+
+<!-- A high level overview of the proposed solution.
+When there are multiple alternatives there should be an explanation
+of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API)
+is likely too low level. -->
+
+Similar to how the Ethereum L1 Consensus-Layer test-vectors (beaconchain only) are generated,
+and more recently the Ethereum L1 Execution Layer test-vectors (EVM only),
+we can piggy-back on a reference implementation, and export the artifacts.
+
+The action tests in the Optimism monorepo can serve as the same kind of testing:
+a collection of interesting state-transitions, based on the defacto reference client (the Go stack),
+easy to export.
+
+The `op-e2e` action tests are called "action" tests because a
+few actors synchronously run through a sequence of actions:
+the test imperatively applies the changes, there is no async scheduling, no clock.
+
+The action tests generally look like:
+- Construct test parameters, configs.
+- Construct a few actors.
+- Get the actors to a starting state, generally at genesis.
+- Apply some actions to actors:
+  - Creating a block as L1 miner, including manual tx inclusion.
+  - Creating a block as L2 sequencer, including manual tx inclusion.
+  - Batch-submitting L2 block(s), queuing up the tx into the L1 tx-pool, to be pulled into L1 when desired.
+  - Proposing an output-root to L1.
+  - Applying finality to L1.
+  - Applying a reorg to L1.
+  - Forwarding the L1 finality signal to a L2 node.
+  - Sharing the latest L1 head signal with the L2 node.
+  - Running the "sync" of a L2 client to the latest signaled head.
+- Assert outcomes about the state.
+
+With some modifications, we could export the data of test-actors during a test-run:
+- Configs of L1 and L2.
+- L1 blocks, receipts, pre/post-state.
+- L2 blocks, receipts, pre/post-state.
+
+These modifications are likely special cases in the L1 and L2 engine actors,
+to export the produced blocks to e.g. JSON files, and any other data such as state.
+
+Potentially, the op-program DB can be reused in the action-tests,
+to capture witness data, such that individual blocks can be executed in a FP setting.
+
+A Fault-Proof program, when given a L1 chain and L2 chain,
+and potentially witness data, should be able to reproduce the state-transition,
+and verify the results match against the expected L2 post-state.
+
+To do more extensive testing, of e.g. alternative L2 consensus nodes,
+we need to also export the reorg signals and safety signals,
+so these offchain concepts can be acted on by the test target.
+
+For interop, ideally the export format includes test-actor names and chain-IDs,
+so we can support multi-chain tests.
+
+### action-tests DSL
+
+The current "actions" DSL has a few quirks, we may want to improve on:
+- Some actor functions produce an action, which then still has to be executed.
```

Comment:
Do you mean method like this https://github.com/ethereum-optimism/optimism/blob/develop/op-e2e/actions/l1_miner.go#L148, which becomes direct execution and does not return the function?